shared/lib_shared_inverter.cpp

```diff
@@ -45,162 +45,161 @@ SharedInverter::SharedInverter(const SharedInverter& orig){
+void SharedInverter::calculateTempDerate(double V, double tempC, double& p_rated, double& ratio, double& loss)
 {
-	if (eff == 0. || pAC == 0.) return;
-
-	double slope = 0.0;
-	double startT = 0.0;
-	double Vdc = 0.0;
-	double slope2 = 0.0;
-	double startT2 = 0.0;
-	double Vdc2 = 0.0;
-
-	// Find the appropriate derate curve depending on DC voltage
-	size_t idx = 0;
-	double deltaT = 0.0;
-	double slopeInterpolated = 0.0;
-	double startTInterpolated = 0.0;
-
-	while (idx < m_thermalDerateCurves.size() && V > m_thermalDerateCurves[idx][0]) {
-		idx++;
-	}
-	if (m_thermalDerateCurves.size() == 1) {
-		Vdc2 = m_thermalDerateCurves[0][0];
-		startTInterpolated = m_thermalDerateCurves[0][1];
-		slopeInterpolated = m_thermalDerateCurves[0][2];
-	}
-	// Use temp and slope of lower and upper curves for interpolation if they both exist
-	else if (idx > 0 && idx < m_thermalDerateCurves.size()) {
-		Vdc2 = m_thermalDerateCurves[idx][0];
-		Vdc = m_thermalDerateCurves[idx - 1][0];
-		double startTGuess = 0.0;
-		double slopeGuess = 0.0;
-		size_t n = std::max(m_thermalDerateCurves[idx].size() / 2, m_thermalDerateCurves[idx - 1].size() / 2);
-		size_t count = 0;
-		while (tempC > startTGuess && count < n ) {
-			findPointOnCurve(idx, startT2, startT2, slope2);
-			findPointOnCurve(idx-1, startT, startT, slope);
-			startTGuess = (startT2 - startT) / (Vdc2 - Vdc)*(V - Vdc2) + startT2;
-			slopeGuess = (slope2 - slope) / (Vdc2 - Vdc)*(V - Vdc2) + slope2;
-			if (tempC > startTGuess) {
-				startTInterpolated = startTGuess;
-				slopeInterpolated = slopeGuess;
-				count++;
-			}
-		}
-	}
-	// otherwise extrapolate using first start temps of each curve in order to avoid inconsistent start temps
-	else {
-		if (idx == 0) {
-			Vdc2 = m_thermalDerateCurves[idx][0];
-			findPointOnCurve(idx, -273, startT2, slope2);
-			Vdc = m_thermalDerateCurves[idx + 1][0];
-			findPointOnCurve(idx + 1, -273, startT, slope);
-			startTInterpolated = (startT2 - startT) / (Vdc2 - Vdc)*(V - Vdc2) + startT2;
-			slopeInterpolated = (slope2 - slope) / (Vdc2 - Vdc)*(V - Vdc2) + slope2;
-		}
-		else {
-			Vdc2 = m_thermalDerateCurves[idx-1][0];
-			findPointOnCurve(idx - 1, -273, startT2, slope2);
-			Vdc = m_thermalDerateCurves[idx - 2][0];
-			findPointOnCurve(idx - 2, -273, startT, slope);
-			startTInterpolated = (startT2 - startT) / (Vdc2 - Vdc)*(V - Vdc2) + startT2;
-			slopeInterpolated = (slope2 - slope) / (Vdc2 - Vdc)*(V - Vdc2) + slope2;
-		}
-	}
-	deltaT = tempC - startTInterpolated;
-
-	// If less than start temp, no derating
-	if (deltaT <= 0) return;
-
-	// If slope is positive, set to zero with no derating
-	if (slopeInterpolated >= 0) return;
-	if (slopeInterpolated < -1) slopeInterpolated = -1;
-
-	// Power in units of W, eff as ratio
-	double pDC = pAC/eff;
-	eff += deltaT* slopeInterpolated;
-	if (eff < 0) eff = 0.;
-	loss = pAC - (pDC * eff);
-	pAC = pDC * eff;
+    if (ratio == 0. || p_rated == 0.) return;
```

Comment:
ratio makes sense. pAC however is the calculated AC power of the inverter. the derate should be happening on the max dc power, so this will have to be renamed.